internal/controller/apiextensions/composite/composed.go

```diff
@@ -159,47 +160,83 @@ func NewAPIConnectionDetailsFetcher(c client.Client) *APIConnectionDetailsFetche
 }
 
 // FetchConnectionDetails of the supplied composed resource, if any.
-func (cdf *APIConnectionDetailsFetcher) FetchConnectionDetails(ctx context.Context, cd resource.Composed, t v1.ComposedTemplate) (managed.ConnectionDetails, error) {
+func (cdf *APIConnectionDetailsFetcher) FetchConnectionDetails(ctx context.Context, cd resource.Composed, t v1.ComposedTemplate) (managed.ConnectionDetails, error) { // nolint:gocyclo
 	sref := cd.GetWriteConnectionSecretToReference()
+	var data map[string][]byte
 	if sref == nil {
-		return nil, nil
+		// We hit this block if the composed resource does not specify a secret ref
+		// this would occur if it does not publish any secrets, however we may want to still
+		// publish some values or access values from fieldPaths.
+		data = make(map[string][]byte)
+	} else {
+		// It's possible that the composed resource does want to write a
+		// connection secret but has not yet. We presume this isn't an issue and
+		// that we'll propagate any connection details during a future
+		// iteration.
+		s := &corev1.Secret{}
+		nn := types.NamespacedName{Namespace: sref.Namespace, Name: sref.Name}
+		if err := cdf.client.Get(ctx, nn, s); client.IgnoreNotFound(err) != nil {
+			return nil, errors.Wrap(err, errGetSecret)
+		}
+		data = s.Data
```

Comment:
What do you think about moving this whole secret-getting block inside the `v1.ConnectionDetailFromConnectionSecretKey` case? It looks like that's the only place that uses `data`, so we could probably just return/continue early if the user asked for a particular connection secret key but there was no connection secret. e.g.:

```go
switch d.Type {
case v1.ConnectionDetailFromConnectionSecretKey:
	sref := cd.GetWriteConnectionSecretToReference()
	if sref == nil {
		continue
	}
	# ...
}
```

Reply:
If I am understanding this correctly, we would need to get the secret each time we loop for every connectionDetail of type `v1.ConnectionDetailFromConnectionSecretKey`, which seems less efficient

Reply:
Ah - good point. I missed the loop.

Another option here would be:

```go
data := map[string][]byte{}
if sref := cd.GetWriteConnectionSecretToReference(); sref != nil {
	s := &corev1.Secret{}
	nn := types.NamespacedName{Namespace: sref.Namespace, Name: sref.Name}
	if err := cdf.client.Get(ctx, nn, s); client.IgnoreNotFound(err) != nil {
		return nil, errors.Wrap(err, errGetSecret)
	}
	data = s.Data
}
```

That said, I'm just nitpicking because I have a weird aversion to `else` statements, so feel free to ignore this suggestion.

Reply:
I like this way more! Looks so much cleaner haha